### PR TITLE
kernel: ath10k: provide a build variant for small RAM devices 

### PR DESCRIPTION
**Pull request by PolynomialDivision**

### CODE REVIEW

**Comment by luaraneda** [comment]
I think some modifications were previously done to ath10k for small RAM devices and were reverted once ath10k-smallbuffers was introduced.
If I'm not wrong they were reverted by 1e27befe63ff4c69f110c7310316f4af75ee63e9

You can compare it with your approach.

**Comment by PolynomialDivision** [comment]
I am doing exactly the same changes, just squashed into a single patch. :)

**Comment by spolack** [comment]
That change is necessary for many wireless mesh networks which use ath10k based devices with too less RAM. The -ct driver has been proven unstable if used with 11s meshing and only wave chipsets are actually supporting 11s.

Freifunk Berlin is nowadays assembling its firmware based completely of vanilla OpenWRT with some package additions which are made through the imagebuilder. Therefore we cannot take the approach other freifunk communities have taken to maintain that patch downstream [1]. Other communities consider these devices as broken and that change would pretty much give those devices a second life. [2]

Thank you in advance for reviewing and hopefully merging! In addition we would really appreciate, if you could backport that change to 21.02 aswell.



¹ https://git.freifunk-franken.de/mirror/openwrt/commit/450b306e540bc0f2c8a8841bbe4d9612f2b8cdea
² https://github.com/freifunk-gluon/gluon/issues/1988#issuecomment-619532909

**Comment by PolynomialDivision** [comment]
@blocktrron @aparcar @nbd168

**Comment by aparcar** [comment]
I'm not fully qualified but if I understand this correctly it adds another kmod which can be used instead of the "regular" buffer one? If it just adds a bit compile and store time, fine with me.

**Comment by aparcar** [comment]
@dangowrt can you have a look?

**Comment by PolynomialDivision** [comment]
> 

Yep. It does the same as ath10k-ct-smallbuffers. The changes are needed otherweise you can not use ath10k und some small ram devices e.g.:
- tp link eap225 outdoor

File: package/kernel/mac80211/Makefile [hunk 1]
```diff
@@ -419,9 +419,15 @@ config-$(call config_package,rsi91x-sdio) += RSI_SDIO
...
+C_DEFINES=
+
+ifeq ($(BUILD_VARIANT),smallbuffers)
+	C_DEFINES+= -DCONFIG_ATH10K_SMALLBUFFERS
+endif
+
 MAKE_OPTS:= -C "$(PKG_BUILD_DIR)" \
 	$(KERNEL_MAKE_FLAGS) \
-	EXTRA_CFLAGS="-I$(PKG_BUILD_DIR)/include $(IREMAP_CFLAGS)" \
+	EXTRA_CFLAGS="-I$(PKG_BUILD_DIR)/include $(IREMAP_CFLAGS) $(C_DEFINES)" \
```
Comment (rsalvaterra) [inline package/kernel/mac80211/Makefile:430]: Do we need the extra variable? Why not simply doing…

```
ifeq ($(BUILD_VARIANT),smallbuffers)
	EXTRA_CFLAGS+= -DCONFIG_ATH10K_SMALLBUFFERS
endif
```

… after the `MAKE_OPTS` definition?

**Review by rsalvaterra — APPROVED** [review]
Other than my previous comment, the rest looks sane to me.

File: package/kernel/mac80211/Makefile (hunk 1 above)
Comment (PolynomialDivision) [reply package/kernel/mac80211/Makefile:430]: If that works, I am happy to do it! :)

Comment (PolynomialDivision) [reply package/kernel/mac80211/Makefile]: @rsalvaterra Did I miss something? For me it seems that the config symbol is not defined then I build mac80211?

Comment (rsalvaterra) [reply package/kernel/mac80211/Makefile]: Oh, silly me! `MAKE_OPTS` is being defined with the complete string. The previous version is fine, sorry about the noise. :/

**Comment by spolack** [comment]
What is blocking this from getting merged? Anything we can do in order to speedup that process? I especially need it on 21.02.

File: package/kernel/mac80211/Makefile (hunk 1 above)
Comment (PolynomialDivision) [reply package/kernel/mac80211/Makefile:430]: Reverted back to previous version. No problem! :)

**Comment by dangowrt** [comment]
Cherry-picked as https://github.com/openwrt/openwrt/commit/694757a08f620a9f24b70003542d9dcd0abeac46 and https://github.com/openwrt/openwrt/commit/236c3ea7306124a6d3aa92e247b7cfd8518bfcc2

**Comment by PolynomialDivision** [comment]
Thanks! :)

**Pull request closed without merging**